Fix "tree scrolls back to left"

nitzan-shaked opened the pull request:
this is a proposal for fixing #1453.
This passes tests for my locally.

along the way:

make Tree._refresh_node somewhat more efficient by:
    * stopping early
    * coallescing multiple refresh() calls of single lines into one call

change "TreeControl" to "Tree" in a few texts
(and funny enough, in demo.css, which doesn't even use it);

remove the (appearantly dead) Tree._refresh_node_line


**Please review the following checklist.**

- [ ] Docstrings on all new or modified functions / classes 
- [ ] Updated documentation
- [ ] Updated CHANGELOG.md (where appropriate)

willmcgugan commented:
There is little benefit in combining multiple refresh calls, since Textual will do this anyway, typically generating a single update.

nitzan-shaked commented:
@willmcgugan yes I get that, and I don't mind reverting that change, if you prefer. It kind of of a natural by-product of the "stop early". To be honest I was a bit lazy to do what I thought would be "the right thing", which is a binary search on the children until you find the one that's just below the view. Or alternatively a binary search up on the nodes in the view to see who isn't "ancestored" by `node`. Both to save the `node in tree_lines[i].path`, which searches in `path` as a sequence. But I decided it probably wasn't worth it.

That way or the other (and do lmk), what do you say about the fix itself?

willmcgugan commented:
I think what we need here is a `get_label_region` on the tree, which returns just the region occupied by the label. Which I think you already calculate. Then we can pass that to `scroll_to_region`.

We can remove the logic to batch updates.

nitzan-shaked commented:
@willmcgugan done: reverted batching logic, and refactored logic into `_get_label_region` which `scroll_to_line()` uses. Passes tests locally. Now running CI tests.

willmcgugan commented on src/textual/widgets/_tree.py:
```diff
@@ -556,6 +556,15 @@ def _get_node(self, line: int) -> TreeNode[TreeDataType] | None:
         else:
             return tree_line.node
 
+    def _get_label_region(self, line: int) -> Region | None:
```
Docstring please.

nitzan-shaked replied:
Oops. Done, thank you.

willmcgugan reviewed (approved):
LGTM
